tracker/internal/graphstore/v1beta/server.go

```diff
@@ -0,0 +1,263 @@
+package v1beta
+
+import (
+	"context"
+	"database/sql"
+	"encoding/base32"
+	"encoding/binary"
+	"time"
+
+	"github.com/depscloud/api/v1beta/graphstore"
+
+	"github.com/sirupsen/logrus"
+)
+
+// GraphStoreServer encapsulates the logic for storing a graph using a generic driver.
+type GraphStoreServer struct {
+	Driver Driver
+}
+
+// Put ...
+func (s *GraphStoreServer) Put(ctx context.Context, request *graphstore.PutRequest) (*graphstore.PutResponse, error) {
+	length := len(request.GetNodes()) + len(request.GetEdges())
+	items := make([]*GraphData, 0, length)
+
+	// add nodes before adding edges
+	// otherwise edges may point to nodes that do not exist
+
+	for _, node := range request.GetNodes() {
+		items = append(items, ConvertNode(node, EncodingProtocolBuffers))
+	}
+
+	for _, edge := range request.GetEdges() {
+		items = append(items, ConvertEdge(edge, EncodingProtocolBuffers))
+	}
+
+	if err := s.Driver.Put(ctx, items); err != nil {
+		return nil, err
+	}
+
+	return &graphstore.PutResponse{}, nil
+}
+
+// Delete ...
+func (s *GraphStoreServer) Delete(ctx context.Context, request *graphstore.DeleteRequest) (*graphstore.DeleteResponse, error) {
+	length := len(request.GetNodes()) + len(request.GetEdges())
+	items := make([]*GraphData, 0, length)
+
+	// delete edges before deleting nodes
+	// otherwise edges may point to nodes that do not exist
+
+	dateDeleted := &sql.NullTime{
+		Time:  time.Now(),
+		Valid: true,
+	}
+
+	for _, edge := range request.GetEdges() {
+		item := ConvertEdge(edge, EncodingProtocolBuffers)
+		item.DateDeleted = dateDeleted
+		items = append(items, item)
+	}
+
+	for _, node := range request.GetNodes() {
+		item := ConvertNode(node, EncodingProtocolBuffers)
+		item.DateDeleted = dateDeleted
+		items = append(items, item)
+	}
+
+	if err := s.Driver.Delete(ctx, items); err != nil {
+		return nil, err
+	}
+
+	return &graphstore.DeleteResponse{}, nil
+}
+
+func paginate(lastToken string, pageSize int) (int, string) {
```

Comment:
Pagination is new in beta? Cool!

Reply:
We had it on list in v1alpha. This changes the semantic a bit.